### DISCUSSION TITLE
How to detect hydration error before production ?

### DISCUSSION THREAD

**Discussion by mrudelle** [nuxt/nuxt#9781] [Questions (Nuxt 2)]
Hi there!

We recently had a problem in production with a hydration error (mismatch between dynamically and statically generated DOM) due to a non-obvious `<p><p></p></p>` caused by a slot tag in a component. It rendered our page unusable for some time.

So we are wondering if there is any way to detect it before production. It seems straightforward but we don't seem to find an answer.

We tried using [nuxt/html-validator](https://html-validator.nuxtjs.org/) which sounds perfect but `generate` actually fixes the issue so html-validator does not detect it.

Any though on how we could detect this kind of error in our CI build script ?

Note: We are serving static pages produced by `nuxt generate` via a s3 bucket.

**Comment by TheAlexLichter** [comment]
As hydration errors only occur during the first request, you could write a simple crawler script that checks for console errors and opens every URL. Besides that, detection ain't that easy 🙁

**Reply by mrudelle** [reply]
Right, detection without rendering seems hard.

Is there a way to tell `nuxt generate` to not fix this kind of problems? So that html-validator catches them afterward and makes the build fail.

Or in a more generic way - it would be useful to make `nuxt generate` fail when it needs to make a change to the DOM structure that will make it diverge from the dynamically created DOM.

**Reply by TheAlexLichter** [reply]
Maybe this is something that could be baked into the Nuxt crawler? 🤔 
On the other hand, you might need a full browser to actually capture the console errors (e.g. using headless Chrome / puppeteer), so this could be tough.

**Reply by mrudelle** [reply]
Thank you for the quick response btw. We'll also setup sentry and see if it catches hydration errors.

Does the crawler yield the list of pages in a machine readable way somewhere? Closest I can think of right now is sitemap.xml.
